File: week1/class2/taylor_series.m
% Clear the workspace and close all figures
clear;
close all;

% Function to approximate the exponential function e^x using a series expansion
function eAprox = eF(x, n=50)
  % Initialize the approximation result
  eAprox = 0;

  % Loop to sum the series up to the n-th term
  for i = 0:n
    % Add the i-th term of the series to the approximation
    eAprox = eAprox + x^i / factorial(i);
  endfor

  return;
endfunction

% Function to approximate the sine function sin(x) using a series expansion
function sinAprox = sinF(x, n=50)
    sinAprox = 0;  % Initialize the approximation result

    % Loop to sum the series up to the n-th term
    for ind=0:n
        % Add the i-th term of the series to the approximation
        sinAprox = sinAprox + (-1)^ind * (x - pi / 2)^(2 * ind) / factorial(2 * ind);
    endfor

    return;
endfunction

% Main script
% Number of iterations for testing the functions
N = 10;

% Loop to compute and display approximations and actual values
for ind = 1:N
    % Calculate and print the approximation and actual value of e^x
    printf('e^%i; aprox=%5.4f, real=%5.4f\n', ind, eF(ind), exp(ind));
    
    % Calculate and print the approximation and actual value of sin(x)
    printf('sen(%5.4f); aprox=%5.4f, real=%5.4f\n\n', ind / N, sinF(ind / N), sin(ind / N));
endfor

File: week1/class2/taylor_series.py
from math import factorial
import math


def e(x: float, n: int = 50) -> float:
    """Approximates the exponential function e^x using a series expansion

    Args:
        x (float): value of x to approximate the exp
        n (int, optional): number of iterations. Defaults to 50.

    Returns:
        float: approximated value of exp
    """
    # Initialize the approximation result
    e_aprox: float = 0.0

    # Loop to sum the series up to the n-th term
    for i in range(n):
        # Add the i-th term of the series to the approximation
        e_aprox += x**i / factorial(i)

    return e_aprox

def sin(x: float, n: int = 50) -> float:
    """Approximate the sine function sin(x) using a series expansion

    Args:
        x (float): value of x to approximate the sine
        n (int, optional): number of iterations. Defaults to 50.

    Returns:
        float: approximated value of sine
    """
    # Initialize the approximation result
    sin_aprox: float = 0.0

    # Loop to sum the series up to the n-th term
    for i in range(n):
        # Add the i-th term of the series to the approximation
        sin_aprox += (-1)**i * (x - math.pi / 2)**(2 * i) / factorial(2 * i)

    return sin_aprox


if __name__ == "__main__":
    # Number of iterations for testing the functions
    N: int = 10
    # Number of decimal places to round the results
    decimals: int = 4

    # Loop to compute and display approximations and actual values
    for i in range(N):
        # Calculate and print the approximation and actual value of e^x
        print('e^{}; aprox={}, real={}'.format(
            i,
            round(e(i), decimals),
            round(math.exp(i), decimals)
        ))

        # Calculate and print the approximation and actual value of sin(x)
        print('sen({}); aprox={}, real={}'.format(
            i / N,
            round(sin(i / N), decimals),
            round(math.sin(i / N), decimals)
        ))
